### PR TITLE
Initial CMake build system

### LINKED ISSUE #74: Port popt to cmake
As rpm will be moving to cmake, we'll want popt there too, starting with popt 1.20. Popt may well be simple enough that we can just do a flag-day swap but that remains to be seen.

### PR DESCRIPTION
**Pull request by Conan-Kudo**
Fixes #74 

- [x] Builds and installs library
- [x] Builds and installs documentation
- [x] Builds and installs translations
- [x] Builds and runs tests
- [x] Builds and runs in CI

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
@pmatilai Any tips on how to get the tests to run? I'm struggling to get them to build with `make test`:

The test executables currently fail to compile:
```
[ 87%] ESC[32mBuilding C object tests/CMakeFiles/test1.dir/test1.c.oESC[0m
In file included from /home/ngompa/Sources/github.com/forks/popt/tests/test1.c:5:
/home/ngompa/Sources/github.com/forks/popt/src/system.h:31:22: error: static declaration of ‘stpcpy’ follows non-static declaration
   31 | static inline char * stpcpy (char *dest, const char * src) {
      |                      ^~~~~~
In file included from /home/ngompa/Sources/github.com/forks/popt/src/system.h:19:
/usr/include/string.h:491:14: note: previous declaration of ‘stpcpy’ with type ‘char *(char * restrict,  const char * restrict)’
  491 | extern char *stpcpy (char *__restrict __dest, const char *__restrict __src)
      |              ^~~~~~
```

**Comment by pmatilai** [comment]
I'll have a look at that test thing. As for translations, I suggest we go the same route as with rpm, and take the opportunity to split them off to a separate popt-l10n effectively managed by Weblate.

**Comment by pmatilai** [comment]
The issue is basically missing config.h. This looks like it's partially based on my cmake draft, I think I tried to just cheapskate around the annoying config.h, but I hadn't gotten to the test-programs yet: the test programs need all the same defines as the main source does, due to relying on system.h. 

So that chkdef() stuff from src/CMakeLists.txt needs to move to the top level cmake file so they get added to test programs as well. Using a config.h[.in] would swipe that cruft from the compiler invocations under the carpet.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+set(OPTINCS
+	fnmatch.h glob.h langinfo.h libintl.h mcheck.h stdalign.h
+)
+foreach(inchdr ${OPTINCS})
+	chkhdr(${inchdr} FALSE)
+endforeach()
+
+
+# Build the library
+set (popt_SOURCES
```
Comment (pmatilai) [inline src/CMakeLists.txt:53]: I know this is how it was in my version and initially in the rpm cmake build as well. I don't find the style *offensive*, but AIUI these sources-as-variables kind of this isn't really in the cmake ethos, and I later changed rpm to just set these in the add_library() call - having them as variables doesn't buy anything at all.

**Comment by Conan-Kudo** [comment]
> I'll have a look at that test thing. As for translations, I suggest we go the same route as with rpm, and take the opportunity to split them off to a separate popt-l10n repo effectively managed by Weblate.

Can we do this split after we merge this in? Doing the conversion in a single repo is much easier...

**Comment by Conan-Kudo** [comment]
I did figure out building and installing translations though, so now it's just tests and strict mode that need to be fixed. 😄

File: src/CMakeLists.txt (hunk 1 above)
Comment (Conan-Kudo) [reply src/CMakeLists.txt:53]: I thought I was going to use it in more places, but now that I don't, I'll merge them into the `add_library()` call.

**Comment by pmatilai** [comment]
Yeah don't worry about the split, I get that you don't want extra moving parts during the conversion. The groundwork of repo setup etc is more or less done now but we'll pull the plug only after cmake is merged.

File: src/CMakeLists.txt (hunk 1 above)
Comment (pmatilai) [reply src/CMakeLists.txt:53]: Yup, coming from automake background, I had similar feelings about it. Turns out it's unnecessary, you can add sources dynamically (not that it's needed for popt) natively in cmake and also get the sources from the target if necessary.

**Comment by pmatilai** [comment]
Just lift the chkdef() and chkhdr() stuff into the main CMakeLists.txt as-is and voila, tests compile and complete (as per my testing yesterday)

**Comment by pmatilai** [comment]
FWIW, I'd leave the different build types with their massive flag lists out of the initial conversion. Just aim for a simple 1:1 build, we can always add more details later but I think just getting the conversion done and over with is the priority right now.

File: src/CMakeLists.txt (hunk 1 above)
Comment (Conan-Kudo) [reply src/CMakeLists.txt]: This is fixed now.

**Comment by Conan-Kudo** [comment]
@pmatilai Now we just have failing tests:

```
ngompa@fedora ~/S/g/f/p/build (cmakeify)> ctest --output-on-failure --force-new-ctest-process
Test project /home/ngompa/Sources/github.com/forks/popt/build
    Start 1: test1_build
1/5 Test #1: test1_build ......................   Passed    0.09 sec
    Start 2: test2_build
2/5 Test #2: test2_build ......................   Passed    0.09 sec
    Start 3: test3_build
3/5 Test #3: test3_build ......................   Passed    0.09 sec
    Start 4: tdict_build
4/5 Test #4: tdict_build ......................   Passed    0.09 sec
    Start 5: testit
5/5 Test #5: testit ...........................***Failed    0.07 sec
Running tests in /home/ngompa/Sources/github.com/forks/popt/build/tests
PASS: test1 - 1
PASS: test1 - 2
PASS: test1 - 3
test1: bad argument --simple: unknown option
FAIL: test1 - 4: "" != "arg1: 0 arg2: another"
test1: bad argument --two: unknown option
FAIL: test1 - 5: "" != "arg1: 1 arg2: alias"
PASS: test1 - 6
test1: bad argument --simple: unknown option
FAIL: test1 - 7: "" != "arg1: 0 arg2: abcd rest: --arg1"
test1: bad argument --takerest: unknown option
FAIL: test1 - 8: "" != "arg1: 1 arg2: (none) rest: --arg2"
PASS: test1 - 9
PASS: test1 - 10
test1: bad argument -T: unknown option
FAIL: test1 - 11: "" != "arg1: 0 arg2: bar"
test1: bad argument -O: unknown option
FAIL: test1 - 12: "" != "arg1: 1 arg2: (none)"
test1: bad argument -OT: unknown option
FAIL: test1 - 13: "" != "arg1: 1 arg2: foo"
PASS: test1 - 14
PASS: test1 - 15
PASS: test1 - 16
PASS: test1 - 17
PASS: test1 - 18
test1: bad argument --echo-args: unknown option
PASS: test1 - 19
test1: bad argument --echo-args: unknown option
FAIL: test1 - 20: "" != "--arg1"
test1: bad argument -T: unknown option
FAIL: test1 - 21: "" != "--arg2 something"
test1: bad argument -T: unknown option
FAIL: test1 - 22: "" != "--arg2 something more args"
test1: bad argument --echo-args: unknown option
FAIL: test1 - 23: "" != "--echo-args -a"
PASS: test1 - 24
PASS: test1 - 25
PASS: test1 - 26
PASS: test1 - 27
PASS: test1 - 28
PASS: test1 - 29
test1: bad argument --grab: unknown option
FAIL: test1 - 30: "" != "arg1: 0 arg2: 'foo bingo' rest: boggle"
test1: bad argument --grabbar: unknown option
FAIL: test1 - 31: "" != "arg1: 0 arg2: 'foo bar' rest: boggle"
PASS: test1 - 32
PASS: test1 - 33
PASS: test1 - 34
PASS: test1 - 35
PASS: test1 - 36
PASS: test1 - 37
PASS: test1 - 38
PASS: test1 - 39
PASS: test1 - 40
PASS: test1 - 41
PASS: test1 - 42
PASS: test1 - 43
PASS: test1 - 44
PASS: test1 - 45
PASS: test1 - 46
PASS: test1 - 47
PASS: test1 - 48
PASS: test1 - 49
PASS: test1 - 50
PASS: test1 - 51
PASS: test1 - 52
PASS: test1 - 53
PASS: test1 - 54
PASS: test1 - 55
PASS: test1 - 56
PASS: test1 - 57
FAIL: test1 - 58: "Usage: test1 [-I?] [-c|--cb2=STRING] [--arg1] [-2|--arg2=ARG]
        [-3|--arg3=ANARG] [-onedash] [--optional=STRING] [--val]
        [-i|--int=INT] [-s|--short=SHORT] [-l|--long=LONG]
        [-L|--longlong=LONGLONG] [-f|--float=FLOAT] [-d|--double=DOUBLE]
        [--randint=INT] [--randshort=SHORT] [--randlong=LONG]
        [--randlonglong=LONGLONG] [--argv=STRING] [--bitset] [--bitclr]
        [-x|--bitxor] [--nstr=STRING] [--lstr=STRING] [-I|--inc]
        [-c|--cb=STRING] [--longopt] [-?|--help] [--usage]" != "Usage: test1 [-I?] [-c|--cb2=STRING] [--arg1] [-2|--arg2=ARG]
        [-3|--arg3=ANARG] [-onedash] [--optional=STRING] [--val]
        [-i|--int=INT] [-s|--short=SHORT] [-l|--long=LONG]
        [-L|--longlong=LONGLONG] [-f|--float=FLOAT] [-d|--double=DOUBLE]
        [--randint=INT] [--randshort=SHORT] [--randlong=LONG]
        [--randlonglong=LONGLONG] [--argv=STRING] [--bitset] [--bitclr]
        [-x|--bitxor] [--nstr=STRING] [--lstr=STRING] [-I|--inc]
        [-c|--cb=STRING] [--longopt] [-?|--help] [--usage] [--simple=ARG]"
FAIL: test1 - 59: "Usage: test1 [OPTION...]
      --arg1                      First argument with a really long
                                  description. After all, we have to test
                                  argument help wrapping somehow, right?
  -2, --arg2=ARG                  Another argument (default: "(none)")
  -3, --arg3=ANARG                A third argument
      -onedash                    POPT_ARGFLAG_ONEDASH: Option takes a single -
      --optional[=STRING]         POPT_ARGFLAG_OPTIONAL: Takes an optional
                                  string argument
      --val                       POPT_ARG_VAL: 125992 141421
  -i, --int=INT                   POPT_ARG_INT: 271828 (default: 271828)
  -s, --short=SHORT               POPT_ARG_SHORT: 4523 (default: 4523)
  -l, --long=LONG                 POPT_ARG_LONG: 738905609 (default: 738905609)
  -L, --longlong=LONGLONG         POPT_ARG_LONGLONG: 738905609 (default:
                                  738905609)
  -f, --float=FLOAT               POPT_ARG_FLOAT: 3.14159 (default: 3.14159)
  -d, --double=DOUBLE             POPT_ARG_DOUBLE: 9.8696 (default: 9.8696)
      --randint=INT               POPT_ARGFLAG_RANDOM: experimental
      --randshort=SHORT           POPT_ARGFLAG_RANDOM: experimental
      --randlong=LONG             POPT_ARGFLAG_RANDOM: experimental
      --randlonglong=LONGLONG     POPT_ARGFLAG_RANDOM: experimental
      --argv STRING               POPT_ARG_ARGV: append string to argv array
                                  (can be used multiple times)
      --[no]bitset                POPT_BIT_SET: |= 0x7777
      --[no]bitclr                POPT_BIT_CLR: &= ~0xf842
  -x, --bitxor                    POPT_ARGFLAG_XOR: ^= (0x8ace^0xfeed)
      --nstr=STRING               POPT_ARG_STRING: (null) (default: null)
      --lstr=STRING               POPT_ARG_STRING: "123456789..." (default:
                                  "This tests default strings and exceeds the
                                  ... limit.
                                  123456789+123456789+123456789+123456789+123456789+ 123456789+123456789+123456789+123456789+123456789+ 1234567...")

arg for cb2
  -c, --cb2=STRING                Test argument callbacks
  -I, --inc                       An included argument

Callback arguments
  -c, --cb=STRING                 Test argument callbacks
      --longopt                   Unused option for help testing

Help options:
  -?, --help                      Show this help message
      --usage                     Display brief usage message" != "Usage: test1 [OPTION...]
      --arg1                      First argument with a really long
                                  description. After all, we have to test
                                  argument help wrapping somehow, right?
  -2, --arg2=ARG                  Another argument (default: "(none)")
  -3, --arg3=ANARG                A third argument
      -onedash                    POPT_ARGFLAG_ONEDASH: Option takes a single -
      --optional[=STRING]         POPT_ARGFLAG_OPTIONAL: Takes an optional
                                  string argument
      --val                       POPT_ARG_VAL: 125992 141421
  -i, --int=INT                   POPT_ARG_INT: 271828 (default: 271828)
  -s, --short=SHORT               POPT_ARG_SHORT: 4523 (default: 4523)
  -l, --long=LONG                 POPT_ARG_LONG: 738905609 (default: 738905609)
  -L, --longlong=LONGLONG         POPT_ARG_LONGLONG: 738905609 (default:
                                  738905609)
  -f, --float=FLOAT               POPT_ARG_FLOAT: 3.14159 (default: 3.14159)
  -d, --double=DOUBLE             POPT_ARG_DOUBLE: 9.8696 (default: 9.8696)
      --randint=INT               POPT_ARGFLAG_RANDOM: experimental
      --randshort=SHORT           POPT_ARGFLAG_RANDOM: experimental
      --randlong=LONG             POPT_ARGFLAG_RANDOM: experimental
      --randlonglong=LONGLONG     POPT_ARGFLAG_RANDOM: experimental
      --argv STRING               POPT_ARG_ARGV: append string to argv array
                                  (can be used multiple times)
      --[no]bitset                POPT_BIT_SET: |= 0x7777
      --[no]bitclr                POPT_BIT_CLR: &= ~0xf842
  -x, --bitxor                    POPT_ARGFLAG_XOR: ^= (0x8ace^0xfeed)
      --nstr=STRING               POPT_ARG_STRING: (null) (default: null)
      --lstr=STRING               POPT_ARG_STRING: "123456789..." (default:
                                  "This tests default strings and exceeds the
                                  ... limit.
                                  123456789+123456789+123456789+123456789+123456789+ 123456789+123456789+123456789+123456789+123456789+ 1234567...")

arg for cb2
  -c, --cb2=STRING                Test argument callbacks
  -I, --inc                       An included argument

Callback arguments
  -c, --cb=STRING                 Test argument callbacks
      --longopt                   Unused option for help testing

Options implemented via popt alias/exec:
      --simple=ARG                simple description

Help options:
  -?, --help                      Show this help message
      --usage                     Display brief usage message"
test1: bad argument --val=foo: option does not take an argument
PASS: test1 - 60
test1: bad argument -x=f1: option does not take an argument
PASS: test1 - 61
PASS: test1 - 62
ln: failed to create symbolic link 'test3-data': File exists
diff: test3-data/01.answer: No such file or directory
FAIL test3 - 1: failed output is in /home/ngompa/Sources/github.com/forks/popt/build/tests/tmp.out, diff is:
diff: test3-data/02.answer: No such file or directory
FAIL test3 - 2: failed output is in /home/ngompa/Sources/github.com/forks/popt/build/tests/tmp.out, diff is:
diff: test3-data/03.answer: No such file or directory
FAIL test3 - 3: failed output is in /home/ngompa/Sources/github.com/forks/popt/build/tests/tmp.out, diff is:

Failed.


80% tests passed, 1 tests failed out of 5

Total Test time (real) =   0.43 sec

The following tests FAILED:
	  5 - testit (Failed)
Errors while running CTest
```

**Comment by Conan-Kudo** [comment]
Oh, it's because it can't find the test data...

**Comment by Conan-Kudo** [comment]
I think we're now good to go!

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	NO_CHECK_REQUIRED_COMPONENTS_MACRO
+)
+install(FILES
+	${CMAKE_BINARY_DIR}/poptConfig.cmake
+	DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/popt
+)
+write_basic_package_version_file(
+	${CMAKE_BINARY_DIR}/poptConfigVersion.cmake
+	VERSION ${PROJECT_VERSION}
+	COMPATIBILITY SameMinorVersion
```
Comment (Conan-Kudo) [inline src/CMakeLists.txt:68]: @pmatilai: One remaining question: do we want to say that we're major version compatible or minor version compatible? I know RPM is minor version compatible, but I feel like popt churns a lot less than RPM does...

Comment (pmatilai) [reply src/CMakeLists.txt:68]: After reading a bit of the docs... Popt is at least *supposed* to be major version compatible. Our just released 1.19 is on the edge but I don't think major version compatible (necessarily) means *bug* compatibility.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+# Build Types
+set(CMAKE_BUILD_TYPE ${CMAKE_BUILD_TYPE}
+	CACHE STRING "Choose the type of build, options are: None Debug Release RelWithDebInfo MinSizeRel SanitizeAddress"
+	FORCE
+)
+
+# AddressSanitize
+set(CMAKE_C_FLAGS_SANITIZEADDRESS
+	"-O2 -g -Wall -fsanitize=address -fno-optimize-sibling-calls -fsanitize-address-use-after-scope -fno-omit-frame-pointer"
```
Comment (pmatilai) [inline CMakeLists.txt:33]: Warning level flags do not belong here. For address sanitation -O2 seems pretty non-optimal in the first place, -fsanitize=address docs explicitly list "-O0, -O1, or -Og" for producing more accurate traces. Moreover, as a developer I want to make my own choices on matters like this, I think address sanitizer option should enable *just that* and not muck with any other compiler flags.

Again I'd leave this out for the initial merge, we can argue the fine details in a separate PR to add this extra functionality.

Comment (Conan-Kudo) [reply CMakeLists.txt]: I forgot to drop the `-O2` actually... But I dropped all of it instead...

File: src/CMakeLists.txt (hunk 2 above)
Comment (Conan-Kudo) [reply src/CMakeLists.txt]: That's what I figured. Adjusted.

**Comment by Conan-Kudo** [comment]
Meh, let's get this in. I don't see any reason to hold it up anymore...

**Comment by pmatilai** [comment]
It would've been nice to wait for an actual ack, you know.

**Comment by pmatilai** [comment]
Anyway, thanks for the effort on this. 

I certainly know the way impatience over getting the almost-there thing in grows the closer you get to the finish line, but when you explicitly ask for a review from somebody then you'll want to give them time to react to latest changes. I certainly don't spend 24/7 on GH.

**Pull request merged**